core/src/main/java/org/elasticsearch/search/suggest/SuggestParseElement.java

```diff
@@ -123,7 +136,85 @@ public SuggestionSearchContext parseInternal(XContentParser parser, MapperServic
+                if (parsefieldMatcher.match(fieldName, TEXT_FIELD)) {
+                    suggestBuilder.setText(parser.text());
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(), "[suggest] does not support [" + fieldName + "]");
+                }
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                if (fieldName == null) {
+                    throw new IllegalArgumentException("Suggestion must have name");
+                }
+                suggestBuilder.addSuggestion(parseSuggestionBuilder(parseContext, fieldName));
+            } else {
+                throw new ParsingException(parser.getTokenLocation(), "[suggest] unknown token [" + token + "] after [" + fieldName + "]");
+            }
+        }
+        return suggestBuilder;
+    }
+
+    public SuggestionBuilder<?> parseSuggestionBuilder(QueryParseContext parseContext, String suggestionName) throws IOException {
+        XContentParser parser = parseContext.parser();
+        ParseFieldMatcher parsefieldMatcher = parseContext.parseFieldMatcher();
+        XContentParser.Token token;
+        String fieldName = null;
+        String suggestText = null;
+        String prefix = null;
+        String regex = null;
+        SuggestionBuilder<?> suggestionBuilder = null;
+
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                fieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if (parsefieldMatcher.match(fieldName, TEXT_FIELD)) {
+                    suggestText = parser.text();
+                } else if (parsefieldMatcher.match(fieldName, PREFIX_FIELD)) {
+                    prefix = parser.text();
+                } else if (parsefieldMatcher.match(fieldName, REGEX_FIELD)) {
+                    regex = parser.text();
+                } else {
+                    throw new IllegalArgumentException("[suggestion] does not support [" + fieldName + "]");
+                }
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                if (suggestionName == null) {
+                    throw new IllegalArgumentException("Suggestion must have name");
+                }
+                Suggester<?> suggester = this.suggesters.get(fieldName);
+                if (suggester == null) {
+                    throw new ParsingException(parser.getTokenLocation(), "Suggestion with name [" + fieldName + "] not supported");
+                }
+                SuggestionBuilder<?> suggestParser = suggester.getBuilderPrototype();
+                if (suggestParser == null) {
+                    throw new IllegalArgumentException("Suggestion with name [" + fieldName + "] not supported");
+                }
+                suggestionBuilder = suggestParser.fromXContent(parseContext, suggestionName);
+            }
+        }
```

Comment:
At this point, it is still possible that the `suggestionBuilder` could be `null`, so it would be better to have a check here to throw an exception if `suggestionBuilder` is `null` here.